text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,171 @@
+config.toml files, but also a copy of each with their default values set.
+
+We moved away from hard-coding defaults in the code a while ago, but a challenge to the other approach is ensuring the
+user _copies_ the installed vanilla config rather than renaming or deleting it.  Two possible solutions are:
+1. have the installer create a read-only copy of the default config exclusively for the `upgrade-bootstrapper`'s use,
+   and a separate writable one for the user to modify/copy/move as they wish
+1. require that every config value is annotated with a standard comment specifying the default value
+
+It seems like annotating the config file with default values might be somewhat more "normal" and should be fairly simple
+to implement and test.
+
+To migrate, the node will identify any modified values in the old config, and apply the same changes to the new config
+if the corresponding new config value has not already been modified by the user.
+
+
+### `upgrade-bootstrapper`
+
+We also need to provide an `upgrade-bootstrapper` binary.  It should live in its own repo outside the `casper-node` one
+if setting up CI is not deemed too onerous.  Ideally its tests shouldn't need to run when changes are made to
+`casper-node`, and vice-versa.  There also should be no direct dependency between the `update-bootstrapper` and any of
+the other casper crates.  Hence it makes sense to keep this separate to the rest of the node project.
+
+The `upgrade-bootstrapper` and `casper-node` must both be capable of being restarted, having been stopped or killed at
+any step above, without disrupting this normal flow.  This will require the `upgrade-bootstrapper` to write which step
+its about to perform to the hard-coded known location immediately before performing the step.  On starting, the
+`upgrade-bootstrapper` will try to read this value and jump to that step.  If the value/file is absent, it assumes it's
+a fresh start.
+
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+Users lose the ability to override config values via command line.
+
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+1. How to get the upgrade activation point agreed by all validators?
+    * Could include first activation point in chainspec, and all chainspec upgrade files hold the _next_ activation
+      point. (Drawback is that we need to know much further in advance when we want the next activation point to occur)
+    * Once new chainspec upgrade file is installed, `upgrade-bootstrapper` could notify the node about its path (via
+      IPC or TCP listener on loopback).  Nodes could then use consensus to agree the point (could take a long time to
+      get consensus as install times of next version of software could vary widely)
+    * On each era change, `casper-node` could read from a file in a known location the next upgrade activation point.
+      This could be the chainspec upgrade file, or a standalone one containing just the activation point
```

Comment:
FWIW, I like this option best. Only I think the upgrade points were supposed to be included in the chainspec, so maybe we should keep them there, and re-read the chainspec at every era change?

Reply:
Well, we'd need to ensure we don't treat _that_ chainspec as current.  It seems like we could reuse the code which reads upgrade points from a file here rather than actually applying the upgrade point to the chainspec.